### PR TITLE
ROX-25547: Add initial Sentinel Alert Notifier

### PR DESCRIPTION
**Pull request by SimonBaeumer**
### Description

This PR adds the Sentinel integration code to send alerts to Sentinel.

 - Send Alert to Sentinel
 - Send Test alerts to Sentinel
 - Introduce a `MarshalMap` function to marshal a proto message into an unstructured map.

### User-facing documentation

- [ ]~~ CHANGELOG is updated **OR** update is not needed~~
- [ ] ~~[documentation PR](https://spaces.redhat.com/display/StackRox/Submitting+a+User+Documentation+Pull+Request) is created and is linked above **OR** is not needed~~

### Testing and quality

- [x] CI results are inspected

#### Automated testing

- [x] added unit tests
- E2E tests are added eventually

#### How I validated my change
  
  - Unit tests are passing
  - Integration sends request to Microsoft Sentinel successfully
 - Export the secrets and run `rox::rest::notifier::sentinel::test` to create a test Alert in Azure Log Analytics workspace

**Scripts I am using:**

```
function rox::notifier::sentinel::config() {
    export SENTINEL_LOG_INGESTION_ENDPOINT=XXXX
    export DIRECTORY_TENANT_ID=XXXX
    export APP_CLIENT_ID=XXXX
    export SECRET=XXXX
    export DATA_COLLECTION_RULE_ID=XXXX
    export STREAM_NAME=XXXX

    logIngestionEndpoint=$SENTINEL_LOG_INGESTION_ENDPOINT
    dataCollectionRuleId=$DATA_COLLECTION_RULE_ID
    streamName=$STREAM_NAME
    directoryTenantID=$DIRECTORY_TENANT_ID
    applicationClientID=$APP_CLIENT_ID
    secret=$SECRET

    notifier=$(cat <<EOF
{
    "name": "sentinel-notifier",
    "uiEndpoint": "localhost:8000",
    "type": "microsoftSentinel",
    "microsoftSentinel": {
        "logIngestionEndpoint": "$logIngestionEndpoint",
        "dataCollectionRuleId": "$dataCollectionRuleId",
        "streamName": "$streamName",
        "directoryTenantId": "$directoryTenantID",
        "applicationClientId": "$applicationClientID",
        "secret": "$secret"
    }
}
EOF
)

    export notifier=$notifier
}

function rox::rest::notifier::sentinel::update() {
    rox::notifier::sentinel::config

    if [[ -z $1 ]]; then
        echo "ID is required."
        return 1
    fi

    roxcurl /v1/notifiers/$1 -v -X PUT --data $notifier -H "Content-Type: application/json"
}

function rox::rest::notifier::sentinel::create() {
    rox::notifier::sentinel::config

    roxcurl /v1/notifiers -v -X POST --data $notifier -H "Content-Type: application/json"
}

function rox::rest::notifier::sentinel::test() {
    rox::notifier::sentinel::config

    roxcurl /v1/notifiers/test -v -X POST --data $notifier -H "Content-Type: application/json"
}
```

### CODE REVIEW

File: proto/storage/proto.lock [hunk 1]
```diff
@@ -11541,7 +11547,7 @@
               },
               {
                 "id": 5,
-                "name": "alert_notifier",
+                "name": "data_collection_rule_config",
```
Comment (dashrews78) [inline proto/storage/proto.lock:11550]: Would probably be best if you made this change in the parent PR before merging it if this is the desired name.

File: proto/storage/notifier.proto [hunk 2]
```diff
@@ -143,12 +144,12 @@ message MicrosoftSentinel {
...
   // TODO(ROX-25739): Add client certificate
 
-  // NotifierConfig contains information about the data collection rule which is a config per notifier type.
+  // DataCollectionRuleConfig contains information about the data collection rule which is a config per notifier type.
   message DataCollectionRuleConfig {
     string stream_name = 1;
     string data_collection_rule_id = 2;
   }
-  DataCollectionRuleConfig alert_notifier = 5;
+  DataCollectionRuleConfig alert_dcr_config = 5;
```
Comment (dashrews78) [inline proto/storage/notifier.proto:152]: see comment on `proto.lock`

File: proto/storage/proto.lock (hunk 1 above)
Comment (SimonBaeumer) [reply proto/storage/proto.lock]: Build broken, the other branch reported the check-generated-files as passing, but actually it failed.

File: central/notifiers/microsoftsentinel/sentinel.go [hunk 3]
```diff
@@ -0,0 +1,162 @@
...
+
+func (s sentinel) AlertNotify(ctx context.Context, alert *storage.Alert) error {
+	if !features.MicrosoftSentinelNotifier.Enabled() {
+		return errors.New("Microsoft Sentinel notifier is disabled.")
+	}
+
+	// convert object to an unstructured map to later wrap it as an array.
+	logToSendObj, err := protocompat.MarshalMap(alert)
+	if err != nil {
+		return errors.Wrap(err, "failed to convert alert to map")
```
[the remark this answers is not in the document]
Comment (vjwilson) [reply central/notifiers/microsoftsentinel/sentinel.go:107]: > If this error gets bubbled up to the UI (I _think_ it does), suggest reword to: "failed to send alert" (the wrapped error will have the gory details)

```suggestion
		return errors.Wrap(err, "failed to send alert to Microsoft Sentinel")
```

Combing through stack-traces is no way to spend one's time. (to paraphrase _Animal House_)

File: central/notifiers/microsoftsentinel/sentinel.go [hunk 4]
```diff
@@ -0,0 +1,162 @@
...
+
+var (
+	log = logging.LoggerForModule(option.EnableAdministrationEvents())
+
+	_ notifiers.AlertNotifier = (*sentinel)(nil)
+)
+
+func init() {
+	if features.MicrosoftSentinelNotifier.Enabled() {
+		log.Info("Microsoft Sentinel notifier enabled.")
```
[the remark this answers is not in the document]
Comment (SimonBaeumer) [reply central/notifiers/microsoftsentinel/sentinel.go:28]: It logs that Sentinel is enabled with a feature flag and adds it to the list of notifiers. If it is not enabled, the notifier is not added.
This is not debug, but protection that customers use it. 
Bumped the log level to debug.

File: central/notifiers/microsoftsentinel/sentinel.go [hunk 5]
```diff
@@ -0,0 +1,162 @@
...
+type sentinel struct {
+	notifier     *storage.Notifier
+	azlogsClient azureLogsClient
+}
+
+func newSentinelNotifier(notifier *storage.Notifier) (*sentinel, error) {
+	config := notifier.GetMicrosoftSentinel()
+
+	// TODO(ROX-25739): Support certificate authentication
+	azureCredentials, err := azidentity.NewClientSecretCredential(config.GetDirectoryTenantId(), config.GetApplicationClientId(), config.GetSecret(), &azidentity.ClientSecretCredentialOptions{})
```
[the remark this answers is not in the document]
Comment (SimonBaeumer) [reply central/notifiers/microsoftsentinel/sentinel.go:44]: Done

File: central/notifiers/microsoftsentinel/sentinel.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (SimonBaeumer) [reply central/notifiers/microsoftsentinel/sentinel.go:107]: Adopted @vjwilson wording

File: central/notifiers/microsoftsentinel/sentinel.go [hunk 6]
```diff
@@ -0,0 +1,162 @@
...
+	logToSendObj, err := protocompat.MarshalMap(alert)
+	if err != nil {
+		return errors.Wrap(err, "failed to convert alert to map")
+	}
+
+	// Wrap object in a slice because Sentinel expects it.
+	logsToSend := []map[string]interface{}{logToSendObj}
+	data, err := json.Marshal(logsToSend)
+	if err != nil {
+		return errors.Wrap(err, "failed to wrap into an array")
```
[the remark this answers is not in the document]
Comment (SimonBaeumer) [reply central/notifiers/microsoftsentinel/sentinel.go:114]: done

File: central/notifiers/microsoftsentinel/sentinel.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (SimonBaeumer) [reply central/notifiers/microsoftsentinel/sentinel.go:28]: CR fixes in this commit on the other PR: https://github.com/stackrox/stackrox/pull/12587/commits/c127178a776a9692aad732457ed4bdb08f8ceba4

File: central/notifiers/microsoftsentinel/sentinel.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (SimonBaeumer) [reply central/notifiers/microsoftsentinel/sentinel.go:44]: CR fixes in this commit on the other PR: https://github.com/stackrox/stackrox/pull/12587/commits/c127178a776a9692aad732457ed4bdb08f8ceba4

File: central/notifiers/microsoftsentinel/sentinel.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (SimonBaeumer) [reply central/notifiers/microsoftsentinel/sentinel.go:107]: CR fixes in this commit on the other PR: https://github.com/stackrox/stackrox/pull/12587/commits/c127178a776a9692aad732457ed4bdb08f8ceba4

File: central/notifiers/microsoftsentinel/sentinel.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (SimonBaeumer) [reply central/notifiers/microsoftsentinel/sentinel.go:114]: CR fixes in this commit on the other PR: https://github.com/stackrox/stackrox/pull/12587/commits/c127178a776a9692aad732457ed4bdb08f8ceba4

**Pull request merged**